### PR TITLE
Parse array type in conditional operator

### PR DESCRIPTION
**Pull request by cston**
### Customer scenario

Parse error reported for array types used in conditional operator expression: `x is T[] ? y : z`

### Bugs this fixes

Internal issue

### Workarounds, if any

Modify code. For instance, add parentheses: `(x is T[]) ? y : z`

### Risk

Low.

### Performance impact

None.

### Is this a regression from a previous update?

Regression from C#7.3.

### How was the bug found?

Customer report

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler, @gafter, please review.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -6262,7 +6245,7 @@ private enum ParseTypeMode
                     while (this.IsPossibleRankAndDimensionSpecifier())
                     {
                         bool unused;
-                        var rank = this.ParseArrayRankSpecifier(mode == ParseTypeMode.ArrayCreation, expectSizes, allowQuestionToken: true, out unused);
+                        var rank = this.ParseArrayRankSpecifier(mode == ParseTypeMode.ArrayCreation, expectSizes, questionTokenModeOpt: mode, out unused);
                         ranks.Add(rank);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: This appears to apply to all levels of a multidimensional array. Shouldn't it apply only to the last?  That is, shouldn't we allow `var z = e is int[]?[] ? x : y;` (existing code)

Comment (cston) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: We allow `e is int[]?[] ? x : y` and `e is int[]?[]? ? x : y`. The issue was we were incorrectly parsing the first case by treating the `?` from the conditional as a nullable qualifier on the array type.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review.
@jaredpar for approval.

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs [hunk 2]
```diff
@@ -291,6 +291,601 @@ public void NullableArray_Cast_05()
...
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "z");
+                }
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void ConditionalOperator_NotNullableArray()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/NullableParsingTests.cs]: >ConditionalOperator_NotNullableArray [](start = 20, length = 36)

Would it be useful to test `x is T?[]` too? I'm afraid we disallow it, but I think we should allow it (seems useful).

**Comment by cston** [comment]
@jaredpar for approval for dev16.0-preview2.

**Comment by jaredpar** [comment]
approved

**Pull request merged**